Add utility for asserting listeners are completed

nicktindall opened the pull request:
Closes #108123

Because we can only be sure that the listener was not called when it finally gets GC'd, it may not fail tests that cause the situation to arise, or even log during them, as the GC may happen some time after they're completed, or in a separate VM. This is somewhat mitigated by the fact that `LeakTracker` will include the test name in the `contextHint`.

Perhaps it would be handy to be able to also trigger these checks in `@AfterEach` (i.e. potentially before GC) so for in-process tests we could fail the test in which the assertion was violated? I have an idea of how this might work

nicktindall committed:
Add utility for asserting listeners are completed

Closes #108123

nicktindall commented on server/src/main/java/org/elasticsearch/action/ActionListenerImplementations.java:
```diff
@@ -24,6 +26,8 @@
  */
 class ActionListenerImplementations {
 
+    private static final Cleaner cleaner = Cleaner.create();
```
The recommendation seemed to be that you have a single `Cleaner` per library. I wonder if we should implement `ElasticSearchCleaner.get()` so that `LeakTracker` and this can share one?

idegtiarenko replied:
It would be really nice to integrate it with a test life-cycle.

nicktindall replied:
I did have a chat with @ywangd about this and he suggested a goal might be to run a node with assertions turned on, so presumably it would nice for it to work in a test lifecycle (trigger assertions at the end of the test that all these handlers created during that test have been completed) but also to kill a node running with assertions turned on when one of these handlers gets garbage collected without being completed (what I think it does now)

nicktindall commented on server/src/main/java/org/elasticsearch/action/ActionListenerImplementations.java:
```diff
@@ -353,4 +357,73 @@ public String toString() {
+                wasCalled = true;
+                delegate.onFailure(e);
+            }
+        }
```
This may be more verbose than it needs to be, but it seemed like the safest pattern from the ones recommended.

https://inside.java/2022/05/25/clean-cleaner/

idegtiarenko replied:
Nice!

nicktindall commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +374,72 @@ public void onFailure(Exception e) {
+                mockLog.assertAllExpectationsMatched();
+            });
+        }
+    }
```
This tests `ActionListener#assertAtLeastOnce`

nicktindall commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +374,72 @@ public void onFailure(Exception e) {
+            assertNotNull(notCalledListener.get());
+            assertNotNull(listenerCreatedAt.get());
+        });
+    }
```
This and the next one test the `ActionListenerImplementations.AssertAtLeastOnceActionListener` directly, as it was easier to do with a specific `NotCalledListener`. Using `MockLog#assertAllExpectationsMatched` seemed brittle when testing that a log message didn't appear.

nicktindall commented on server/src/main/java/org/elasticsearch/action/ActionListenerImplementations.java:
```diff
@@ -353,4 +357,84 @@ public String toString() {
+        @Override
+        public String toString() {
+            return delegate.toString();
+        }
```
Consistent with `assertOnce`, makes the wrapper invisible if someone calls `toString`

nicktindall commented on server/src/main/java/org/elasticsearch/action/ActionListener.java:
```diff
@@ -425,6 +425,28 @@ public boolean equals(Object obj) {
+                        "Expected listener "
+                            + delegate
+                            + " to be called at least once, but it was never called. Created:"
+                            + ExceptionsHelper.formatStackTrace(createdAt.getStackTrace())
```
The logging in `maybeDieOnAnotherThread` strips off the cause, so I put the created-at stack trace in the exception message. I may be better to set it as the cause of the `AssertionError` but I don't know what that will look like being handled by the uncaught exception handler.

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

nicktindall committed:
Fix formatting

nicktindall committed:
Remove unnecessary escaping

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/ActionListenerImplementations.java:
```diff
@@ -353,4 +357,84 @@ public String toString() {
+
+            private CallTrackingActionListener(ActionListener<R> delegate, NotCalledListener listener) {
+                this.delegate = delegate;
+                this.created = new ElasticsearchException("Listener created at...");
```
```suggestion
                this.created = new ElasticsearchException("Listener was never called");
```

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/ActionListenerImplementations.java:
```diff
@@ -353,4 +357,84 @@ public String toString() {
+         */
+        AssertAtLeastOnceActionListener(ActionListener<Response> delegate, NotCalledListener listener) {
+            super(new CallTrackingActionListener<>(delegate, listener));
+            cleaner.register(this, (Runnable) this.delegate);
```
https://openjdk.org/jeps/447 could help us to remove this cast in the future

idegtiarenko commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +374,75 @@ public void onFailure(Exception e) {
+            )
+        );
+        // Call onResponse and/or onFailure at least once
+        IntStream.range(0, 1 + randomInt(3)).forEach(i -> {
+            if (randomBoolean()) {
+                listenerRef.get().onResponse("succeeded");
+            } else {
+                listenerRef.get().onFailure(new RuntimeException("Failed"));
+            }
+        });
```
I do not think it is expected that a listener is called more than once

nicktindall replied:
OK that's good to know. I considered that it might be something that happens, due to the presence of e.g. `notifyOnce` etc. Happy to remove the loop if we think it's noise?

nicktindall replied:
removed! thanks

idegtiarenko commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +374,75 @@ public void onFailure(Exception e) {
+        });
+        assertBusy(() -> {
+            // Nullify reference so it becomes unreachable
+            listenerRef.set(null);
+            System.gc();
```
I wonder if this is going to be stable on ci? How long does it usually take for the reference to be collected?

idegtiarenko replied:
I tired locally and it was reliably calling gc for me completing the test in < 100ms.
I wonder if we should log every attempt to call `System.gc()` here? I am not sure it will be working the same way on various jdks running all tests

nicktindall replied:
Yes it seems unreliable, hopefully in practice it is not. Something to watch.

nicktindall committed:
Update server/src/main/java/org/elasticsearch/action/ActionListenerImplementations.java

Co-authored-by: Ievgen Degtiarenko <ievgen.degtiarenko@gmail.com>

pxsalehi commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +374,75 @@ public void onFailure(Exception e) {
+            );
+            assertBusy(() -> {
+                // Nullify reference so it becomes unreachable
+                listenerRef.set(null);
```
Does setting the ref to null repeatedly make any difference?

nicktindall replied:
Nah I guess it was just to separate the creation of the reference and clearing of the reference into separate blocks.

nicktindall replied:
I moved it out to make it clearer what might need repeating

pxsalehi commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +374,75 @@ public void onFailure(Exception e) {
+                listenerRef.get().onFailure(new RuntimeException("Failed"));
+            }
+        });
+        assertBusy(() -> {
```
Doesn't `ensureUnreachable` make sure that `gc()` is called and tries for 10s? Why is the `assertBusy` necessary?

nicktindall replied:
Good point, that can be moved out (it was just adapted from the previous test)

nicktindall committed:
Remove/minimise assertBusy blocks

pxsalehi commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -434,13 +434,11 @@ public void testAssertAtLeastOnceWillNotLogWhenResolvedOrFailed() throws Excepti
-        });
+        // Nullify reference so it becomes unreachable
+        listenerRef.set(null);
+        System.gc();
```
I was thinking even this System.gc() is not needed.

nicktindall replied:
Good call, removed

nicktindall committed:
Remove unnecessary System.gc()

nicktindall committed:
Merge branch 'main' into fix/108123_add_utility_assert_listeners_complete

nicktindall committed:
Only test for a single call to onResponse OR onFailure

nicktindall committed:
Merge branch 'main' into fix/108123_add_utility_assert_listeners_complete

DaveCTurner reviewed (changes requested):
I wonder if we need so much extra infrastructure here, vs say using the `LeakTracker` directly:

```java
static <Response> ActionListener<Response> assertAtLeastOnce(ActionListener<Response> delegate) {
    if (Assertions.ENABLED) {
        return ActionListener.runBefore(delegate, LeakTracker.INSTANCE.track(delegate)::run);
    } else {
        return delegate;
    }
}
```

Certainly not so easy to test so thoroughly, but then again such a simple change doesn't need such comprehensive testing either.

That said, the leak tracker itself doesn't have such good test coverage today. I'd rather we fixed that rather than introducing something that's effectively a different implementation of the same thing with better tests.

nicktindall committed:
Use LeakTracker to detect non-completed ActionListeners

nicktindall committed:
Merge branch 'main' into fix/108123_add_utility_assert_listeners_complete

nicktindall committed:
fix whitespace

nicktindall commented:
> Certainly not so easy to test so thoroughly, but then again such a simple change doesn't need such comprehensive testing either.
> 
> That said, the leak tracker itself doesn't have such good test coverage today. I'd rather we fixed that rather than introducing something that's effectively a different implementation of the same thing with better tests.

This is a good point, I updated the PR to use `LeakTracker` directly. It actually helps with associating a leak with the test that caused it, because the test infrastructure already adds the test name in `LeakTracker#contextHint`.

I didn't really beef up the testing of `LeakTracker` at all, but added `ESTestCase#assertLeakDetected` which allows a test to assert that a leak was detected, and prevent failing because of it. That might be useful for writing more comprehensive `LeakTracker` tests? Happy to go ahead and do that if we think it's a good approach?

nicktindall commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +370,40 @@ public void onFailure(Exception e) {
+        final AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+            ActionListener.assertAtLeastOnce(ActionListener.running(() -> {
+                // Do nothing, but don't use ActionListener.noop() as it'll never be garbage collected
+            }))
```
This is perhaps a shortcoming, if the listener is something that is retained in some other way (such as the listener returned by `noop()`) the detection won't work because it'll never become unreachable. I'm not sure how common this is though. We could insert another pass-through delegate in the chain that isn't referenced from anywhere to make it more robust, but that would allocated an additional object for what might be a rare edge-case.

nicktindall committed:
Test that assertAtLeastOnce delegates correctly

nicktindall committed:
Simplify regex

nicktindall committed:
Remove variable

nicktindall commented on server/src/main/java/org/elasticsearch/action/ActionListener.java:
```diff
@@ -425,6 +426,16 @@ public boolean equals(Object obj) {
+     */
+    static <Response> ActionListener<Response> assertAtLeastOnce(ActionListener<Response> delegate) {
+        if (Assertions.ENABLED) {
+            return ActionListener.runBefore(delegate, LeakTracker.INSTANCE.track(delegate)::close);
```
This actually ends up being `assertExactlyOnce` with this implementation because `runBefore` invokes `assertOnce`. I wonder if we should call the method `assertExactlyOnce`.

DaveCTurner replied:
Ah good point, well spotted. Let's fix that so that this method returns a listener can be called at least once (but still using `LeakTracker.INSTANCE.track(delegate)`).

nicktindall committed:
Merge branch 'main' into fix/108123_add_utility_assert_listeners_complete

ywangd commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +372,57 @@ public void onFailure(Exception e) {
 
+    public void testAssertAtLeastOnceWillLogAssertionErrorWhenNotResolved() throws Exception {
+        assumeTrue("assertAtLeastOnce will be a no-op when assertions are disabled", Assertions.ENABLED);
+        final AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
```
I guess a plain mutable local variable should work as well?

nicktindall committed:
Tidy tests, add tests for LeakTracker

nicktindall commented on server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:
```diff
@@ -0,0 +1,193 @@
+        void closeLeak();
+    }
+
+    private static class PojoTrackedObjectTestCase implements TrackedObjectTestCase {
```
naming of static classes was dictated by the gradle testTestConventions task

DaveCTurner replied:
Hm I don't think we enforce this naming convention, it's just noise in IntelliJ. I would rather name these things better I think.

nicktindall commented on server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:
```diff
@@ -0,0 +1,193 @@
+    /**
+     * Encapsulates the lifecycle for a particular type of tracked object
+     */
+    public interface TrackedObjectTestCase {
```
Probably more elaborate than I would have liked, but the lifecycles of the different wrappers seemed different enough to make it worthwhile.

nicktindall committed:
Tidy up

nicktindall commented on server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:
```diff
@@ -0,0 +1,206 @@
+
+        private static Releasable createReleasable() {
+            int number = Randomness.get().nextInt();
+            return () -> logger.info("Prevents this returning a non-collectible constant {}", number);
```
I originally returned `() -> {}` but for some reason (I think the compiler turned it into a constant) it would never get garbage collected that way. Adding the random parameter is one way of ensuring that doesn't happen.

DaveCTurner replied:
Yeah `() -> {}` is a static constant. I think `Releasables.assertOnce(()->{})` would work tho.

nicktindall committed:
Remove unnecessary code

nicktindall committed:
De-verbose-ify

nicktindall committed:
Merge branch 'main' into fix/108123_add_utility_assert_listeners_complete

DaveCTurner commented on server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:
```diff
@@ -0,0 +1,161 @@
+     */
+    public interface TrackedObjectTestCase {
+
+        default void makeAssumptions() {};
```
Let's just suppress all these tests if `Assertions.ENABLED == false` I think, no need to be this discriminating here.

DaveCTurner commented on server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java:
```diff
@@ -0,0 +1,161 @@
+        }
+
+        private static RefCounted createRefCounted() {
+            return new AbstractRefCounted() {
```
Could this be `AbstractRefCounted.of(()->{})`?

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -788,6 +789,18 @@ protected static void checkStaticState() throws Exception {
+     */
+    protected static void assertLeakDetected(String expectedPattern) {
+        synchronized (loggedLeaks) {
+            assertTrue(loggedLeaks.removeIf(leakText -> Pattern.matches(expectedPattern, leakText)));
```
Neat.

DaveCTurner reviewed (commented):
Looks good, left some more comments. I suggest breaking out the leak-tracker tests into a separate PR too.

nicktindall committed:
Don't use #runBefore because it changes semantics to exactly-once

nicktindall committed:
LeakTrackerTest improvements

nicktindall committed:
Delete LeakTrackerTests (to introduce in a separate PR)

nicktindall committed:
Add message to assertLeakDetected

nicktindall commented:
> Looks good, left some more comments. I suggest breaking out the leak-tracker tests into a separate PR too.

Comments are addressed now, I've applied the suggested changes and stored the LeakTracker locally ready to open a separate PR after this is merged (it depends on `assertLeakDetected`.

DaveCTurner commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +372,53 @@ public void onFailure(Exception e) {
+        ActionListener<Object> listenerRef = reachabilityChecker.register(ActionListener.assertAtLeastOnce(ActionListener.running(() -> {
+            // Do nothing, but don't use ActionListener.noop() as it'll never be garbage collected
+        })));
+        // Call onResponse or onFailure
+        if (randomBoolean()) {
+            listenerRef.onResponse("succeeded");
+        } else {
+            listenerRef.onFailure(new RuntimeException("Failed"));
+        }
```
I know an earlier iteration called the listener multiple times here and we said to remove that, but in fact I think we should call it multiple times to make sure another `assertOnce` doesn't creep back in. Sorry for the back-and-forth on this point :)

nicktindall replied:
Yep no worries, I nearly added that back after I made the change away from `runBefore`

DaveCTurner commented on server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:
```diff
@@ -370,6 +372,53 @@ public void onFailure(Exception e) {
+        reachabilityChecker.ensureUnreachable();
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testAssertAtLeastOnceWillDelegateResponses() {
+        ActionListener<Object> delegate = mock(ActionListener.class);
+        ActionListener<Object> listener = ActionListener.assertAtLeastOnce(delegate);
+        listener.onResponse("succeeded");
+        verify(delegate).onResponse("succeeded");
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testAssertAtLeastOnceWillDelegateFailures() {
+        ActionListener<Object> delegate = mock(ActionListener.class);
+        ActionListener<Object> listener = ActionListener.assertAtLeastOnce(delegate);
+        RuntimeException exception = new RuntimeException();
+        listener.onFailure(exception);
+        verify(delegate).onFailure(exception);
+    }
```
Could we do this without mocks? Mocks tend to get in the way of future refactorings, best avoided IMO unless there really is no other choice. Something like this perhaps?

```suggestion
    public void testAssertAtLeastOnceWillDelegateResponses() {
        final var response = new Object();
        assertSame(
            response,
            safeAwait(SubscribableListener.newForked(l -> ActionListener.assertAtLeastOnce(l).onResponse(response)))
        );
    }

    public void testAssertAtLeastOnceWillDelegateFailures() {
        final var exception = new RuntimeException();
        assertSame(
            exception,
            safeAwaitFailure(SubscribableListener.newForked(l -> ActionListener.assertAtLeastOnce(l).onFailure(exception)))
        );
    }
```

DaveCTurner reviewed (commented):
Looks good, just a couple of nits on the tests.

nicktindall committed:
Update server/src/test/java/org/elasticsearch/action/ActionListenerTests.java

Co-authored-by: David Turner <david.turner@elastic.co>

nicktindall committed:
Review feedback

nicktindall committed:
Merge branch 'main' into fix/108123_add_utility_assert_listeners_complete

DaveCTurner reviewed (approved):
LGTM
